Weblink - Allow relative urls in config

ReneNulschDE opened the pull request:
## Description:
Allow absolute URLs in configuration

**Related issue (if applicable):** fixes #11077

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4465

## Example entry for `configuration.yaml` (if applicable):
```yaml
weblink:
  entities:
    - name: Grafana
      url: /grafana
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

ReneNulschDE committed:
Allow relative url

ReneNulschDE committed:
Allow absolute urls in config schema

ReneNulschDE committed:
change after pylint build

balloob commented on homeassistant/components/weblink.py:
```diff
@@ -20,8 +20,9 @@
-    # pylint: disable=no-value-for-parameter
-    vol.Required(CONF_URL): vol.Url(),
+    vol.Required(CONF_URL): vol.Any(
+        vol.Match(r'\A/[a-z0-9]', msg='invalid absolut url'),
```
"Invalid relative URL"

balloob replied:
It's also not matching all relative URLs. `/states/group.bla` would fail.

ReneNulschDE replied:
thank you for your review. I'll change the text. The second one is interesting. ```/states/group.bla``` works at runtime but the test case fails.

balloob reviewed (changes requested):
Please add a test.

ReneNulschDE committed:
Add tests and change error message

ReneNulschDE commented:
@balloob I've added the tests and also changed the error message. In my test the URL `/states/group.bla`  worked with the regex `'\A/[a-z0-9]'`. Could you give me some more advice on this point?

balloob commented:
Ah, it looks like it passes indeed.

However, the regex is very weird, you only match the first character after the `/`, why is that ?

![image](https://user-images.githubusercontent.com/1444314/35191526-06a60d04-fe32-11e7-8261-ed9501b34d37.png)

ReneNulschDE commented:
my intent was: Allow an absolute path only, otherwise, the user could get strange effects in the frontend. "My" definition of an absolute path was: It needs a starting forward slash and then at least one character. If anybody has a better approach, I'll be happy to change this PR.

balloob commented:
I think that testing that it starts with a `/` is all the validation we should do.

MartinHjelmare commented:
Why is the title of this PR `Allow relative` when the description says `Allow absolute`. It's confusing. Reading the conversion and the docs PR it sounds as if only absolute url will be allowed. Please clarify, thanks!

balloob commented:
I changed the title to relative because that's what is being added. An absolute url starts with `http://`. Looks like the constant names and error messages are all wrong indeed.

ReneNulschDE committed:
Change regex to check starting forward slash only

ReneNulschDE commented:
I have changed the error message to "Invalid absolute path in relative URL".

ReneNulschDE committed:
Change error message and const name
